crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -166,6 +172,60 @@ impl Client {
             }
         }
     }
+
+    pub fn transaction_stream(
+        self,
+        block: BlockNumber,
+    ) -> impl futures::Stream<Item = PeerData<Vec<Transaction>>> {
+        async_stream::stream! {
+            // Loop which refreshes peer set once we exhaust it.
+            loop {
+                let peers = self
+                    .get_update_peers_with_sync_capability(protocol::Transactions::NAME)
+                    .await;
+
+                // Attempt each peer.
+                'next_peer: for peer in peers {
+                    let request = TransactionsRequest {
+                        iteration: Iteration {
+                            start: block.get().into(),
+                            direction: Direction::Forward,
+                            limit: 1,
```

Comment:
There is an additional provision in the p2p spec which allows interleaving of block data (step). This can be used to stream concurrently from multiple sources without large batches. e.g. open a stream from peer A which download blocks `[0,2,4,6,...]` and another from peer B which does blocks `[1,3,5,..]`.

A downside to only requesting one block at a time is that we may get rate-limited or blocked out by the peer. We need to sync millions of blocks eventually, doing so block-per-request seems not ideal.

I left concurrency out entirely for the initial header sync implementation. I just left some comments about where we could introduce concurrency/parallel workers. Headers are of course a bit different since we care more about order - we need to verify the integrity of the parent -> child chain, so there is more reason to keep things in-order. Concurrency for headers can be introduced